### PR TITLE
JS Streamer and Frontend Tests

### PR DESCRIPTION
**Pull request by mcottontensor**
JS Streamer in Extras/JSStreamer
Frontend tests in Extras/FrontendTests

### CODE REVIEW

File: Extras/FrontendTests/tests/extras.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    command: string;
+};
+
+// a generic type for inputs captured by the streamer
+export type DataChannelEvent = DataChannelMouseInput | DataChannelKeyboardInput | DataChannelCommandInput;
+
+// sets up the streamer page to capture data channel messages
+export function setupEventCapture(streamerPage: Page) {
+    return streamerPage.evaluate(() => {
+        window.data_messages = {};
```
Comment (lukehb) [inline Extras/FrontendTests/tests/extras.ts]: Maybe small comment here about attaching this js object and the listener to populate it to the window as (I imagine) this is the most convenient place that is visible to both caller and callee.

File: Extras/FrontendTests/tests/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+        const video_aspect_ratio = this.video_size.height / this.video_size.width;
+        this.player_is_larger = player_aspect_ratio > video_aspect_ratio;
+        if (this.player_is_larger) {
+            this.ratio = player_aspect_ratio / video_aspect_ratio;
+        } else {
+            this.ratio = video_aspect_ratio / player_aspect_ratio;
+        }
+    }
+
+    public normalizeQuantizeSigned(x: number, y: number): Coord {
```
Comment (lukehb) [inline Extras/FrontendTests/tests/helpers.ts:95]: Could be potential to move somethings into common here and have both frontend and extra use it?

File: Extras/FrontendTests/tests/keyboard.spec.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+test('Test keyboard events', {
+    tag: ['@keyboard'],
+}, async ({ page, streamerPage, streamerId }) => {
+
+    await page.goto(`/?StreamerId=${streamerId}&MatchViewportRes=true`);
+    await page.getByText('Click to start').click();
+    await helpers.waitForVideo(page);
+
+    const player_box = await page.locator('#videoElementParent').boundingBox();
```
Comment (lukehb) [inline Extras/FrontendTests/tests/keyboard.spec.ts]: Any particular reason for `underscored_variables` - seems a departure from the rest of the codebase?

File: Extras/FrontendTests/tests/keyboard.spec.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                expected_actions.push({ type: PSEventTypes.KeyDown, key_code: c });
+                expected_actions.push({ type: PSEventTypes.KeyPress, key_code: c });
+                expected_actions.push({ type: PSEventTypes.KeyUp, key_code: c });
+            }
+        });
+
+        // check we got the expected events
+        const first_player_id = Object.keys(events)[0];
+        const single_player_events = events[first_player_id];
+        expect(single_player_events).toContainActions(expected_actions);
```
Comment (lukehb) [inline Extras/FrontendTests/tests/keyboard.spec.ts]: That's nice!

File: Extras/FrontendTests/tests/helpers.ts (hunk 2 above)
Comment (mcottontensor) [reply Extras/FrontendTests/tests/helpers.ts:95]: Yeah I was thinking the same thing as I was doing this. I feel like the input stuff on frontend could be cleaned up a lot.

File: Extras/FrontendTests/tests/keyboard.spec.ts (hunk 3 above)
Comment (mcottontensor) [reply Extras/FrontendTests/tests/keyboard.spec.ts]: It was picked up from somewhere. Maybe when I started doing playwright tests, and was trying to be consistent and it became a habit when moving into this code base.
The joys of trying to be consistent with everyone.

**Review by lukehb — APPROVED** [review]

**Pull request merged**